Add mimx8ml8 m7 support

trowbridgec1 opened the pull request:
Hello!

This PR adds support for the Cortex M7 core of the NXP i.MX 8M Plus SoC as well as the [NXP i.MX8M Plus EVK board](https://www.nxp.com/design/development-boards/i-mx-evaluation-and-development-boards/evaluation-kit-for-the-i-mx-8m-plus-applications-processor:8MPLUSLPD4-EVK) using the existing support for the [NXP i.MX8M Mini EVK board](https://www.nxp.com/design/development-boards/i-mx-evaluation-and-development-boards/evaluation-kit-for-the-i-mx-8m-mini-applications-processor:8MMINILPD4-EVK) (which is already present: https://docs.zephyrproject.org/latest/boards/arm/mimx8mm_evk/doc/index.html) as a template.

This PR is also dependent on this PR (~~https://github.com/zephyrproject-rtos/hal_nxp/pull/125~~ https://github.com/zephyrproject-rtos/hal_nxp/pull/123) to the NXP HAL module.

Note: this is my first time adding a PR for this project, so please let me know if I'm missing anything!

trowbridgec-laird commented:
After discussing with @mmahadevan108, I've closed my PR to the NXP HAL repo (https://github.com/zephyrproject-rtos/hal_nxp/pull/125) in favor of depending on PR https://github.com/zephyrproject-rtos/hal_nxp/pull/123 instead.

trowbridgec-laird commented:
Latest commit addresses style errors.

trowbridgec-laird commented:
Latest commit addresses **more** style issues. Apparently, addressing the warnings is not optional. :slightly_smiling_face:

danieldegrasse commented on soc/arm/nxp_imx/mimx8ml8_m7/soc.c:
```diff
@@ -0,0 +1,383 @@
+	}
+
+	/* Disable MPU */
+	ARM_MPU_Disable();
```
MPU regions in other SOCs are defined in a separate file, as Zephyr has an architecture level MPU driver (in `arch` folder). See this file of an example of how custom MPU regions are defined: https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/st_stm32/stm32h7/mpu_regions.c

trowbridgec-laird replied:
Thank you for the hint! See the new `mpu_regions.c` file.

danieldegrasse commented on soc/arm/nxp_imx/mimx8ml8_m7/soc.c:
```diff
@@ -0,0 +1,383 @@
+	   should be multiples of region size, to make it simple, the MPU region 5 set the address
+	   space 0x80000000 ~ 0xBFFFFFFF to be non-cacheable. Then MPU region 6 set the text and
+	   data section to be cacheable if the program running on DDR. The cacheable area base
+	   address should be multiples of its size in linker file, they can be modified per your
```
Zephyr has a `nocache` section that is linked to the memory region defined as `SRAM` by the linker files. Could you select DDR memory as the SRAM region, or does your non cacheable region need to be defined in a separate memory region from general RAM?

trowbridgec-laird replied:
@danieldegrasse - This part is a little tricky as it's an auxiliary core to the Cortex A cores in the i.MX8M Plus SoC. The documentation and SDK from NXP shows that the M7 core supports running from several different regions, and they provide separate example linker files for each.

From what I can tell, the M7 core supports using the `nocache` `SRAM` section provided by Zephyr (the DTCM) in addition to a second `nocache` `SRAM` section in the DDR region.

One important caveat is that the M7 core supports running entirely from the DDR, and in this case, the `FLASH` section must be marked as cacheable.

I'll play with the settings here and see if I can come up with something more modular to support the different modes.

trowbridgec-laird replied:
See the latest changes. I moved to using the Zephyr MPU driver and added a `mpu_regions.c` file to match the same configuration.

I also added a section to the linker script for the additional DDR region that is available.

danieldegrasse replied:
Just checked, those changes look good, thanks for adding them! See my comment below, to keep the commit history of Zephyr clean those changes need to be squashed into the prior commit.

danieldegrasse commented on boards/arm/mimx8mp_evk/mimx8mp_evk.yaml:
```diff
@@ -0,0 +1,20 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+identifier: mimx8mp_evk
```
Can you add a `supported` definition for UART, to enable the CI to test UART drivers?

trowbridgec-laird replied:
Done.

trowbridgec-laird commented on soc/arm/nxp_imx/mimx8ml8_m7/soc.c:
```diff
@@ -0,0 +1,383 @@
+
+	CLOCK_EnableClock(kCLOCK_Rdc);   /* Enable RDC clock */
+	CLOCK_EnableClock(kCLOCK_Ocram); /* Enable Ocram clock */
+	CLOCK_EnableClock(kCLOCK_Audio); /* Enable Audio clock to power on the audiomix domain*/
```
You're correct! Zephyr doesn't yet have an `AUDIOMIX` driver, so I just removed the audio initialization code for now as that's out of scope of this PR (adding initial support for the core).

danieldegrasse commented on soc/arm/nxp_imx/mimx8ml8_m7/soc.c:
```diff
@@ -321,22 +106,19 @@ static void SOC_ClockInit(void)
-	/* switch AUDIO AHB to SYSTEM PLL1 */
-	CLOCK_SetRootMux(kCLOCK_RootAudioAhb, kCLOCK_AudioAhbRootmuxSysPll1);
-
+#ifdef CONFIG_UART_MCUX_IUART
```
Since it looks like this clock is specific to UART4, you may want to add to this `#ifdef` to check that the uart4 node is enabled in the DTS. See here for an example: https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/nxp_imx/rt/soc_rt10xx.c#L198

trowbridgec-laird replied:
done! See latest update.

danieldegrasse reviewed (commented):
Can you squash the commits that remove the AUDIOMIX peripheral and change the MPU region definition into the original commit? This will help keep the change history of Zephyr clean.

trowbridgec-laird commented:
@danieldegrasse Please see the latest update with squashed commits!

mbolivar-nordic commented on boards/arm/mimx8mp_evk/mimx8mp_evk_itcm.dts:
```diff
@@ -0,0 +1,32 @@
+	compatible = "nxp,mimx8mp_evk";
+
+	aliases {
+		uart-4 = &uart4;
```
These types of aliases are generally not needed in new boards. They occur in old boards as a holdover from the days before we had things like DT_NODELABEL. Do you really need this?

trowbridgec-laird replied:
Removed!

mbolivar-nordic commented on boards/arm/mimx8mp_evk/mimx8mp_evk_ddr.dts:
```diff
@@ -0,0 +1,32 @@
+	compatible = "nxp,mimx8mp_evk";
+
+	aliases {
+		uart-4 = &uart4;
```
Same question here

trowbridgec-laird replied:
Removed!

trowbridgec-laird commented:
The HAL module PR dependency for this PR merged (https://github.com/zephyrproject-rtos/hal_nxp/pull/123), so I rebased this branch on `main`.

trowbridgec-laird commented:
Latest push updates the `hal_nxp` dependency in `west.yml` to the latest revision.

dleach02 commented:
@trowbridgec-laird, the HAL PR actually merged this morning. Can you rebase so we can let this PR go through CI?

trowbridgec-laird commented:
@dleach02 Rebased and pushed my changes!

trowbridgec1 committed:
boards: arm: Add NXP i.MX8M Plus EVK board support

Add board support for NXP i.MX8M Plus EVK. This board has the following
features:

Processor    : i.MX8M Plus Quad applications processor
Memory       : 32-bit LPDDR4 w/6 GB
               eMMC 5.0/5.1 w/32 GB
               SD/MMC connector
               QSPI w/32 MB
Connectivity : MIMO 2x2 Wi-Fi 802.11b/g/n/ac and BT 4.2
               2x Ethernet (1x w/ TSN)
               PCIe M.2
               2x CAN FD DB9 Female connectors
USB          : USB 3.0 Type C for Power
               USB 3.0 Type A
               USB 3.0 Type C
Debug        : JTAG connector
               MicroUSB for debug console

More information about this board can be found in NXP website: https://www.nxp.com/design/development-boards/i-mx-evaluation-and-development-boards/evaluation-kit-for-the-i-mx-8m-plus-applications-processor:8MPLUSLPD4-EVK

Signed-off-by: Chris Trowbridge <chris.trowbridge@lairdconnect.com>

trowbridgec-laird commented:
Latest commit adds `#include <devicetree.h>` to `soc.h` for this SoC since some Zephyr architecture code (e.g., `thread.c`) utilizes the devicetree helpers and macros.

trowbridgec-laird commented:
Just checking in - can this PR be merged?
